drucker/drucker_dashboard_servicer.py

```diff
@@ -149,27 +154,35 @@ def SwitchModel(self,
         return drucker_pb2.ModelResponse(status=1,
                                          message='Success: Switching model file.')
 
+    @error_handling(drucker_pb2.EvaluateModelResponse(
+        metrics=drucker_pb2.EvaluationMetrics(num=0, accuracy=0.0, precision=[0.0], recall=[0.0], fvalue=[0.0], option={})))
     def EvaluateModel(self,
                       request_iterator: Iterator[drucker_pb2.EvaluateModelRequest],
                       context: _Context
                       ) -> drucker_pb2.EvaluateModelResponse:
-        """ Evaluate your ML model.
-        :TODO: in detail.
+        """ Evaluate your ML model and save result.
         """
-        try:
-            for evaluateModelRequest in request_iterator:
-                test_data = evaluateModelRequest.data
-                result = self.app.evaluate(test_data)
-                return drucker_pb2.EvaluateModelResponse(num=result.num,
-                                                         accuracy=result.accuracy,
-                                                         precision=result.precision,
-                                                         recall=result.recall,
-                                                         fvalue=result.fvalue)
-        except Exception as e:
-            self.logger.error(str(e))
-            self.logger.error(traceback.format_exc())
-            return drucker_pb2.EvaluateModelResponse(num=0,
-                                                     accuracy=0,
-                                                     precision=0,
-                                                     recall=0,
-                                                     fvalue=0)
+        first_req = next(request_iterator)
+        save_path = first_req.data_path
+
+        test_data = b''.join([first_req.data] + [r.data for r in request_iterator])
+        result, details = self.app.evaluate(test_data)
+        metrics = drucker_pb2.EvaluationMetrics(num=result.num,
+                                                accuracy=result.accuracy,
+                                                precision=result.precision,
+                                                recall=result.recall,
+                                                fvalue=result.fvalue,
+                                                option=result.option)
+        tmp_path = self.app.get_eval_path(uuid.uuid4().hex)
+        def write_and_move(postfix, data):
+            path = tmp_path + postfix
+            Path(path).parent.mkdir(parents=True, exist_ok=True)
+            with open(path, 'wb') as f:
+                pickle.dump(data, f)
+            eval_path = self.app.get_eval_path(save_path + postfix)
+            Path(eval_path).parent.mkdir(parents=True, exist_ok=True)
+            shutil.move(path, eval_path)
+        write_and_move(self.EVALUATE_RESULT, result)
+        write_and_move(self.EVALUATE_DETAIL, details)
```

Comment:
Do not you save the evaluation data itself?

Reply:
EvaluateDetail includes input and label data, which is evaluation data.
So I think it's good enough

Reply:
But when we run the test for another `service`, do we need to send the same data again?

Reply:
@keigohtr 
Umm, that's right.
So we should like below?

1. frondend-dashboard sends hash value of test data file to backend-dashboard
2. backend-dashboard checks the hash
    - If the value is stored in db
        1. send the hash value instead of data itself to DruckerDashboardServicer
        2. DruckerDashboardServicer get the saved test data from the hash value
        3. run evaluation and return the evaluate result
    - else
        1. backend-dashboard returns false or something to frondend-dashboard
        2. frondend-dashboard sends the test data to backend-dashboard
        3. backend-dashboard sends the test data to DruckerDashboardServicer
        4. DruckerDashboardServicer gets evaluate result
        5. DruckerDashboardServicer saves the test data (and evaluate result)
        6. DruckerDashboardServicer returns the hash value of the test data (and evaluate result) to backend-dashboard
        7. backend-dashboard saves the hash value to db and return evaluate result to frondend-dashboard

Reply:
@yuki-mt Umm, seems complicated.

At the moment, I think all we need is `EVALUATE_DETAIL` but excluding the evaluation data. We don't need to save `EVALUATE_RESULT` also.

For the future, the component we need are...
- [ ] Save the evaluation data to manage the data itself.
- [ ] Use the saved evaluation data when user request to run an evaluation.
  - [ ] Save the result avoiding the duplicated execution.

Reply:
@yuki-mt I understood what you mean.
I think your current spec is good to me.

We need to modify this spec as soon as possible by the way. This spec has a potential to consume the disk when the evaluation data is huge.

Reply:
It is also related to `EvaluateResult` , so I will change the spec in the next PR